Latin cube 2

ablaom opened the pull request:
PR #83 did not integrate new code and tests with the rest of the package. (They are islands.) Also, some of these tests are actually failing. 

- [x] Add `include` statements to bring in code and tests
- [x] Resolve failed tests

@ludoro Can you please fix the tests?

ludoro commented:
I have rewritten the test set to make it much more comparable to the other ones (e.g DummyModel and SuperModel). 
The function _create_bounds_and_dims works fine, see last test case. However I believe there is some problem with how  `n`  is used in the code. `n` in this case number of points to have in the plan, passed in by the user when he calls TunedModel().
I believe the problem is at line 85 of the src/strategies/latin_hypercube, the n there should be model.n? I am not really sure.
Now that I think about it, should I also define the default_N myself?

ablaom commented:
Okay, you want to add a parameter `max_iterations` or `goal` to your LatinHypercube struct and in setup use `tuning.max_iterations` for `n` in `setup`.

ablaom commented:
Oops. Corrected previous comment

ablaom commented:
@ludoro How about you set `default_n(tuning, range) = tuning.n_max`? Then if the user fails to specify `n` in his `TunedModel`, he used `n_max` specified in the `LatinHypercube` struct.

ablaom commented on Project.toml:
```diff
@@ -33,7 +33,7 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 MultivariateStats = "6f286f6a-111f-5878-ab1e-185364afe411"
-NearestNeighbors = "b8a86587-4115-5ab1-83bc-aa920d37bbce"
+#NearestNeighbors = "b8a86587-4115-5ab1-83bc-aa920d37bbce"
 StableRNGs = "860ef19b-820b-49d6-a774-d7a799459cd3"
```
Not sure you can put comments in the toml file this way.

Are you getting a pkg conflict? These can be tricky to sort out and I'd be glad to take a look.

ludoro replied:
Yeah, not sure why. At the moment the CI thing seems to be stuck, tomorrow if this fail I will just revert back to an old commit in that pile that I know works.

ablaom commented:
Yeah, put NearestNeighbors back into [extras]. The merge conflict has occurred for unrelated reason. StaticArrays has released a 1.0, which is causing widespread havoc. In our case we need NearestNeighbors to update [compat] to StaticArrays 1.0. I'll make a PR there now.

ablaom commented:
So we are waiting for https://github.com/KristofferC/NearestNeighbors.jl/pull/114

ludoro commented:
Oh wow, thanks a lot. I was really wondering how I could have messed up so badly, ahah. 
NearestNeighbors is back into extras.

ablaom commented:
After https://github.com/JuliaRegistries/General/pull/25681 is merged tray closing and re-opening the issue or making a trivial commit to re-trigger tests. Note that 1.5 is failing silently (it reports the pkg conflict but ignores it). You might have to retry hours later - it seems github actions is slow to update General or something.

ablaom commented:
@ludoro Tests are passing. You ready for  final review?

ludoro commented:
Not yet, we are still missing one more test (number 3) for the setup function which I believe will fail due to some problem with the library. Let me push that and see.

ludoro commented:
So, I solved many issues up until this one, which happens in the last function I need to test. So I'd guess we are mostly there.
I get an error when `setup` calls `makeLatinHypercube` which calls` recursive_setproperty!`, which gives an error related to Int(2.6) not being precise. The only place in my code where I convert a float to an integer is in `src/strategies/latin_hypercube.jl` on line 105, where I call `r[j].values[Int(scaled_plan[i,j])]`, I do that because I need to convert my 1.0,2.0,3.0 etc to actual hyperparamter values contained in `r[j].values` (e.g they are characters), of course I do this only when the `dims[j]` is Categorical, as the check on line 100 requires. So it is supposed to be called only when the number being approximated is e.g Int(1.0) -> 1, so I don't get why that happens in `recursive_setproperty!`. What's more is that this particular call should not happen in this specific test that is failing because the two ranges are actually numeric ranges, so the Int(float) thing should never be called in the "setup" test. All the other tests are fine.
I guess I need some help understanding how to go about fixing this issue with recursive set property because I do not know what to touch.

ablaom commented:
The problem is that you are trying to set integer fields of the model to non-integer values (both `Dummy.Model.alpha` and `DummyModle.lambda` have type `Int`) but you have generated the values using the Latinhypercube algorithm, as if you were preparing to mutate float fields. So there is actually a design decision to be made here: For *discrete* but *numeric* fields, how should the strategy generate the final values? For consistency with grid, I would say that you do precisely what you are already doing, but round to the nearest integer before mutating (calling `recursive_setproperty!`). 

Currently you do not seem to pass the field types to `makeLatinHypercube`, only the `dims` vector, which basically tracks if it is numeric (aka `LatinHypercubeSampling.Continuous`) or nominal. You actually need a finer distinction here, so you know when to round. It would be messy to pass both `dims` and the types, since the types already determines `dims`. So I suggest you pass a `range_types` vector or tuple and use dispatch to get the `dims` object. So, something like:

```julia
_latin_type(::Type) = LatinHypercubeSampling.Categorical # fallback
_latin_type(::Type{<:Real}) = LatinHypercubeSampling.Continuous
_range_type(::MLJBase.ParamRange{T}) were T = T
...
range_types = _range_type.(r) # where `r` is a vector of one-d ranges

# and when you need it:
dims = _latin_type.(range_types)
```

You should probably add unit tests for these dispatchers to make sure they do what you (or I 😄 ) expect.

Does this make sense?

ludoro commented:
I see, just to re iterate to see if I understand what's missing: it could be the case that in the scaled_plan matrix in the setup function one value is 2.6 because the dims is continuous, however it should actually be an integer depending on its `r[j]` that is passed. So inside the loop in lines 98-108, I can just add a condition that checks for this and rounds the number to an `Int`, right? 
In this case, I do not even need to pass anything else to the `makeLatinHypercube` function because the scaled plan is already right at that point. (containing both float , integers and possibly even chars) depending on the type of dims.
If that's right, I do not see the need of defining the function `_latin_type()` to accomplish this, is there something I am missing? 

I agree that `dims` looks a bit generic, does `dims_type` look better to you? At the end of the day it is the type of the dimension, either continuous or categorical. I am open to suggestions though!

ablaom commented:
Yes, you understand perfectly and your alternative solution sounds good to me!

ludoro commented:
I think we are ready for final review 😄

ablaom commented:
Since we defer to the LatinHypercubeSampling library for docs, hyperparameters of the sampling strategy that come from the  library should have identical names. So please make these changes:

-  `popSize` -> `popsize`
-  `nTournament` -> `ntour`
-  `pTournament` -> `ptour`
-  `nGenerations` -> `gens`

(BTW I think camelcase for julia variables is not common. So `MyModule` and `MyType` but `my_variable`.)

ablaom commented on src/strategies/latin_hypercube.jl:
```diff
@@ -75,32 +75,34 @@ function _create_bounds_and_dims(d,r)
         for j = 1:size(scaled_plan,2)
             if dims[j] isa LatinHypercubeSampling.Continuous
-                scaled_plan[i,j] = inverse_transform(MLJBase.Scale,
+                if r[j] isa MLJBase.NumericRange{Int,MLJBase.Bounded,Symbol}
```
This is flawed. The type could be any subtype of `Integer` not just `Int=Int64`.  Please don't use `Int(floor...)` use `round(T, ...)` where `T <: Integer` is the type. Rather than use `if` statements, more Julian to use dispatch (more likely to be type-stable). You can see examples of this idea in my earlier comment.

ablaom commented:
@ludoro I have added plotting to LatinHypercube to help with testing. 

I discovered a bug in `models(::LatinHypercube,...)` which I fixed. Probably should have a unit test for `models` (where the bug was) and we l have no integration test. The integration test should create a `TunedModel` instance, wrap it with some data in a machine, and this machine should be `fit!`. For example, see the test at https://github.com/alan-turing-institute/MLJTuning.jl/blob/ed437afba3bb0a54f2c119577b31f101e88a9905/test/strategies/grid.jl#L129 .

The whole `n_max` thing is pretty ugly (first you set `n_max` and then you set `n` in tuned model instance, but there is no reason for these to be different). However, to fix this requires a breaking change. I'll see if I can make a PR for this soon.

ablaom commented:
@ludoro Looks like you missed some changes in the tests. 

You may want to consider the following unix command line tools (work on my mac). You should execute them at the top level of your MLJTuning directory:

```
find . -name "*.jl" -exec grep -nH "nGenerations" {} \; 
```
will print out all occurrences of "nGenerations".  When you are sure you want to replace all of those, then do:

```
find . -name "*.jl" -exec sed -i .backup s/nGenerations/gens/g {} \;
```

to make the substitutions globally in your directory. A bunch of backup files with extension .backup will be created. You can get rid of these with

```
find . -name "*.backup" -delete
```
Apply usual cautions with use of git.

ludoro commented:
I added the test for the models as you suggested and also the _round function as we talked about in the last call before Christmas. Can you check if that is what you had in mind?

ablaom commented:
@ludoro I've finished up with the details. Many thanks for your contribution! 🎉 

Will make those API changes at #106 before tagging a new release.
